src/ThingooConnector.cpp

```diff
@@ -0,0 +1,77 @@
+/***************************************************************
+   ThingooConnector is a library for the ESP8266/Arduino platform
+   to communicate with Thingoo API
+   Built for KNI Project "Thingoo" https://github.com/ThingooKNI/
+ **************************************************************/
+#include "ThingooConnector.h"
+
+ThingooConnector::ThingooConnector(const char *host)
+{
+	_host = _clean_end_point(String(host));
+}
+
+void ThingooConnector::set_client_credentials(String client_id, String secret_key)
+{
+	_client_id = client_id;
+	_secret_key = secret_key;
+}
+
+void ThingooConnector::set_ssl_certificate_fingerprint(const char* fingerprint)
+{
+	_fingerprint = fingerprint;
+}
+
+void ThingooConnector::connect()
+{
+	String token = _get_token();
+	Serial.println(token);
+}
+
+String ThingooConnector::_get_token()
+{
+	std::unique_ptr<BearSSL::WiFiClientSecure > client(new BearSSL::WiFiClientSecure);
+	client->setFingerprint(_fingerprint);
+	String token_endpoint = "https://" + (String) _host + register_endpoint;
+	http.begin(*client, token_endpoint);
+	http.addHeader("Content-Type", "application/x-www-form-urlencoded");
+	String http_request_data = "grant_type=client_credentials&client_id=" + _client_id + "&client_secret=" + _secret_key;
+
+	int http_response_code = http.POST(http_request_data);	//Send HTTP POST request
+
+	if (http_response_code == 200)
+	{
+		String payload = http.getString();
+		DeserializationError error = deserializeJson(doc, payload);
+
+		if (error)
+		{
+			Serial.print(F("deserializeJson() failed: "));
+			Serial.println(error.f_str());
+		}
+
+		String access_token = doc["access_token"];
+		return access_token;
+	}
+	else
+	{
+		String error_message = "[HTTP] POST... failed, error: " + String(http_response_code);
```

Comment:
Don't return a String. Throw own exception instead. Create struct/class extending std::exception with some proper name, AccessTokenRetrievalException for example. Here's an article on how to do that: http://peterforgacs.github.io/2017/06/25/Custom-C-Exceptions-For-Beginners/

Reply:
I've done that but not with success. I had to add 2 lines of code in platformio.ini:
build_flags = -fexceptions
build_unflags = -fno-exceptions
Reason? Default Arduino framework doesn't handling exceptions. I tried struct and class with std::exception, but everything was end with:

User exception (panic/abort/assert)
--------------- CUT HERE FOR EXCEPTION DECODER ---------------

Abort called

>>>stack>>>

ctx: cont
sp: 3ffffb20 end: 3fffffc0 offset: 0000
3ffffb20:  00000000 00000000 00000000 00000000  
3ffffb30:  000000fe 00000000 00000000 00000000  
3ffffb40:  00000000 00000000 00000000 b8b1aabc  
3ffffb50:  3ffef77c 3fff3244 00000000 3ffef69c  
3ffffb60:  3ffeec3c 000000c8 3fff3274 40213b0a  
3ffffb70:  00000000 00000000 00000000 40213b1c  
3ffffb80:  00000000 3ffffc1c 00000000 40213481  
3ffffb90:  3ffffc14 00000000 00000000 00000000  
3ffffba0:  00000000 00000000 00000000 40225a30  
3ffffbb0:  00000000 3ffef69c 3ffeec3c 40225a41  
3ffffbc0:  3fff3274 3fff3244 3fff3274 4022626c  
3ffffbd0:  00000000 00000000 3fff3274 4021348c  
3ffffbe0:  40000000 00000000 00000000 00000000  
3ffffbf0:  00000000 00000000 00000000 40225a30  
3ffffc00:  3fff329c 3ffef69c 3ffeec3c 40225a41  
3ffffc10:  3fff3274 402261a0 3fff3274 40226217  
3ffffc20:  3ffeec3c 000000c8 3ffeec38 40210802  
3ffffc30:  3fff3200 0069006f 80746365 38333120  
3ffffc40:  00000000 3ffffc90 3ffffc70 3ffffc60  
3ffffc50:  00000009 0000002e 3fff12dc 402145e5  
3ffffc60:  3ffe8796 00000000 3c000a0d 40214641  
3ffffc70:  00001860 00000003 0000000a 4021156a  
3ffffc80:  000017c8 3ffffcd0 3ffffcd0 00000000  
3ffffc90:  3ffef728 00000030 00000020 40100a42  
3ffffca0:  3ffef728 00000030 00000020 40100a42  
3ffffcb0:  00000001 3ffffd20 00000000 40100a89  
3ffffcc0:  00000001 3ffffd20 00000000 40100a89  
3ffffcd0:  205d3100 3ffe8f00 80fffd90 40200000  
3ffffce0:  205d3100 800c0000 80003234 3fff1400  
3ffffcf0:  0021002f 80000030 3fff329c 0069006f  
3ffffd00:  00fe865a 3fff1400 000c000f 80212162  
3ffffd10:  3fff1500 004a004f 80fffd00 3fff1400  
3ffffd20:  0010001f 80000339 3fff17dc 004a004f  
3ffffd30:  00ff148c 3fff170c 3ffef51c 3ffef69c  
3ffffd40:  3ffffd70 000001f0 00000020 40100a0b  
3ffffd50:  00000024 00000024 3ffef51c 3ffef69c  
3ffffd60:  000001f0 000001f0 3ffeec38 40210915  
3ffffd70:  000001f0 3fffff80 3ffeec38 402120d4  
3ffffd80:  3fffdad0 000001f0 3ffeec38 402010c9  
3ffffd90:  00000000 00000000 3fff147c 3fff1481  
3ffffda0:  3fff1484 3fff1414 3fff1424 3fff1424  
3ffffdb0:  3fff142c 3fff1440 3fff1440 3fff1444  
3ffffdc0:  3fff1458 3fff1458 4021b9b8 00000000  
3ffffdd0:  4021b9b8 00000000 4021b9b8 00000000  
3ffffde0:  4021b9b8 00000000 4021b9b8 00000000  
3ffffdf0:  4021b9b8 00000000 4021b9b8 00000000  
3ffffe00:  feef5035 6e2d6f6e fe007465 86efeffe  
3ffffe10:  feefef00 feefeffe 80efeffe feefef00  
3ffffe20:  feefeffe 80efeffe feefef00 feefeffe  
3ffffe30:  80efeffe feefef00 feefeffe 80efeffe  
3ffffe40:  feefef00 feefeffe 80efeffe 00000000  
3ffffe50:  00000000 00000000 00000000 00000000  
3ffffe60:  00000001 00505345 feefeffe 83efeffe  
3ffffe70:  feefef03 00000000 00000000 00000000  
3ffffe80:  000007d0 00010000 00000000 0050ef00  
3ffffe90:  feef0701 ffffffff 00000000 00000000  
3ffffea0:  01000001 00010101 01010001 fe010001  
3ffffeb0:  3ffe8f7f 3ffe8f7f feefef00 feefeffe  
3ffffec0:  80efeffe 3fff145c 000b000f 00efeffe  
3ffffed0:  feef0000 00007530 feefef00 feefef00  
3ffffee0:  feefeffe 80efeffe feef0100 00000000  
3ffffef0:  feef0100 00000000 00000000 00000005  
3fffff00:  00000000 feefef01 3a6d772a feefef00  
3fffff10:  84efeffe 00000002 feefef02 3ffef56c  
3fffff20:  feefeffe feefeffe 00000000 feefeffe  
3fffff30:  feefeffe feefeffe 00000000 feefeffe  
3fffff40:  feefeffe feefeffe 00000000 feefeffe  
3fffff50:  feefeffe feefeffe 00000000 feefeffe  
3fffff60:  feefeffe feefeffe 00000000 feefeffe  
3fffff70:  feefeffe feefeffe 00000000 feefeffe  
3fffff80:  3fff1400 0024002f 80efeffe 3fff1600  
3fffff90:  000e000f 80efeffe feefeffe feefeffe  
3fffffa0:  3fffdad0 00000000 3ffef65c 402136b8  
3fffffb0:  feefeffe feefeffe 3ffe850c 40100d05  
<<<stack<<<

Of course, only when I passed double trailing slash in host in main.cpp to call out this error handling. 
When I passed host properly - it was going well.
I don't know, maybe its ok exception...? Rather no. I have been stuck with it since tomorrow. Maybe you have some idea to deal with it?

Reply:
https://community.platformio.org/t/exception-handling-disabled-use-fexceptions-to-enable/8762
About exception handling in Arduino